File: day3.py
#!/usr/bin/python3
from shared import read_from_file

tree_map = read_from_file('inputs/day3.txt')

def count_trees(tree_map, directions):
  x_pos = 0
  y_pos = 0
  move_right, move_down = directions
  tree_count = 0
  while y_pos < len(tree_map):
    if tree_map[y_pos][x_pos % len(tree_map[y_pos])] == "#":
      tree_count += 1
    y_pos += move_down
    x_pos += move_right
  return tree_count

count_multiple = [
  [1, 1],
  [3, 1],
  [5, 1],
  [7, 1],
  [1, 2]
]

res = 1
for count in count_multiple:
  res *= count_trees(tree_map, count)

print("Part 1", count_trees(tree_map, [3, 1]))
print("Part 2", res)

File: node/day3.js
const tools = require('./tools')
const tree_map = tools.readFromFile('../inputs/day3.txt')

function traverse_map(tree_map, move) {
  let [move_right, move_down] = move
  let x_pos = 0
  let y_pos = 0
  let tree_count = 0
  while (y_pos < tree_map.length) {
    if (tree_map[y_pos][x_pos % tree_map[y_pos].length] === "#") {
      tree_count++
    }
    y_pos += move_down
    x_pos += move_right
  }

  return tree_count
}

function count_trees(tree_map, input_directions) {
  let result = 1
  input_directions.forEach((directions) => {
    let tree_count = traverse_map(tree_map, directions)
    result *= tree_count
  })

  return result
}

input_a = [
  [3,1]
]
input_b = [
  [1, 1],
  [3, 1],
  [5, 1],
  [7, 1],
  [1, 2]
]

console.log(`Part 1: ${count_trees(tree_map, input_a)}`)
console.log(`Part 2: ${count_trees(tree_map, input_b)}`)